File: src/ruby/builder_pattern_1.rb

class KidsMealBuilder 
  def build_drink
  end

  def build_main
  end

  def build_dessert
  end

  def get_meal
  end
end

class AdultMealBuilder 
  def build_drink
  end

  def build_main
  end

  def build_dessert
  end

  def get_meal
  end
end

class MealDirector
  def create_meal(meal_builder)
    builder.build_drink
    builder.build_main
    builder.build_dessert
    builder.get_meal
  end
end

directory = MealDirector.new
builder = nil
if true
  builder = KidsMealBuilder.new
else
  builder = AdultMealBuilder.new
end
director.create_meal(builder)

File: src/std/lava/general/patterns/BuilderPattern1.java
package std.lava.general.patterns;
class Meal {
}

abstract class MealBuilder {
    protected Meal meal = new Meal();
    public abstract void buildDrink();
    public abstract void buildMain();
    public abstract void buildDessert();
    public abstract Meal getMeal();
}

class KidsMealBuilder extends MealBuilder {
    public void buildDrink() {
        // add drinks to the meal
    }
    
    public void buildMain() {
        // add main part of the meal
    }

    public void buildDessert() {
        // add desert part to the meal
    }

    public Meal getMeal() {
        return meal;
    }
}

class AdultMealBuilder extends MealBuilder {
    public void buildDrink() {
        // add drinks to the meal
    }
    
    public void buildMain() {
        // add main part of the meal
    }

    public void buildDessert() {
        // add desert part to the meal
    }

    public Meal getMeal() {
        return meal;
    }
}

class MealDirector {
    public Meal createMeal(MealBuilder builder) {
        builder.buildDrink();
        builder.buildMain();
        builder.buildDessert();
        return builder.getMeal();
    }
}

public class BuilderPattern1 {
    public static void main(String[] args) {
        MealDirector director = new MealDirector();
        boolean isKid = true;
        MealBuilder builder = null;
        if (isKid) {
            builder = new KidsMealBuilder();
        }
        else {
            builder = new AdultMealBuilder();
        }
        Meal meal = director.createMeal(builder);
    }
}
